src/execution/IncrementalPublisher.ts

```diff
@@ -85,237 +84,401 @@ export type FormattedIncrementalResult<
   | FormattedIncrementalDeferResult<TData, TExtensions>
   | FormattedIncrementalStreamResult<TData, TExtensions>;
 
-export function yieldSubsequentPayloads(
-  subsequentPayloads: Set<IncrementalDataRecord>,
-): AsyncGenerator<SubsequentIncrementalExecutionResult, void, void> {
-  let isDone = false;
+/**
+ * This class is used to publish incremental results to the client, enabling semi-concurrent
+ * execution while preserving result order.
+ *
+ * The internal publishing state is manages as follows:
+ *
+ * '_released': the set of Incremental Data records that are ready to be sent to the client,
+ * i.e. their parents have completed and they have also completed.
+ *
+ * `_pending`: the set of Incremental Data records that are definitely pending, i.e. their
+ * parents have completed so that they can no longer be filtered. This includes all Incremental
+ * Data records in `released`, as well as Incremental Data records that have not yet completed.
+ *
+ * `initialResult`: a record containing the state of the initial result, as follows:
+ * `isCompleted`: indicates whether the initial result has completed.
+ * `children`: the set of Incremental Data records that can be be published when the initial
+ * result is completed.
+ *
+ * Each Incremental Data record also contains similar metadata, i.e. these records also contain
+ * similar `isCompleted` and `children` properties.
+ *
+ * @internal
+ */
+export class IncrementalPublisher {
+  initialResult: {
+    children: Set<IncrementalDataRecord>;
+    isCompleted: boolean;
+  };
 
-  async function next(): Promise<
-    IteratorResult<SubsequentIncrementalExecutionResult, void>
-  > {
-    if (isDone) {
-      return { value: undefined, done: true };
-    }
+  _released: Set<IncrementalDataRecord>;
+  _pending: Set<IncrementalDataRecord>;
 
-    await Promise.race(Array.from(subsequentPayloads).map((p) => p.promise));
+  // these are assigned within the Promise executor called synchronously within the constructor
+  _signalled!: Promise<unknown>;
+  _resolve!: () => void;
 
-    if (isDone) {
-      // a different call to next has exhausted all payloads
-      return { value: undefined, done: true };
-    }
+  constructor() {
+    this.initialResult = {
+      children: new Set(),
+      isCompleted: false,
+    };
+    this._released = new Set();
+    this._pending = new Set();
+    this._reset();
+  }
 
-    const incremental = getCompletedIncrementalResults(subsequentPayloads);
-    const hasNext = subsequentPayloads.size > 0;
+  _trigger() {
```

Comment:
Methods also should be `private`.
BTW. If you know how to configure ESLint to check this it will be great.

Reply:
not sure about how to configure eslint for this...

...but, all good for this file [8ea9e99](https://github.com/graphql/graphql-js/pull/3894/commits/8ea9e990db37de4d631a8a63516425ce7df3bd00)